### PR TITLE
drivers/i3c: support i3c driver frameworks

### PR DESCRIPTION
**Pull request by Donny9**
## Summary
drivers/i3c: support i3c driver frameworks

The I3C (Improved Inter-Integrated Circuit) driver is a comprehensive software framework designed to support the advanced features of the I3C protocol. It comprises a system framework and underlying IP drivers, both of which are critical components for enabling seamless communication between I3C devices. The system framework, provided by vela os, offers user-level device nodes and kernel-level driver function interfaces. Meanwhile, the underlying IP drivers, developed specifically for the I3C driver framework, handle the implementation of IP functionalities and the adaptation of framework-level interfaces.

## Impact
support driver frameworks: i3c
## Testing
Vela

### CODE REVIEW

**Review by jerpelea — APPROVED** [review]
please add the SPDX identifier 

 * SPDX-License-Identifier: Apache-2.0
 * 
to the license

**Review by xiaoxiang781216 — APPROVED** [review]

**Comment by Donny9** [comment]
> please add the SPDX identifier
> 
> * SPDX-License-Identifier: Apache-2.0
> 
> to the license

Done~

**Review by acassis — APPROVED** [review]

File: drivers/i3c/device.c [hunk 1]
```diff
@@ -0,0 +1,346 @@
...
+      nxmutex_unlock(&dev->desc->ibi_lock);
+    }
+
+  i3c_bus_normaluse_unlock(dev->bus);
+
+  return ret;
+}
+
+/****************************************************************************
+ * Name: i3c_device_enable_ibi()
```
Comment (pkarashchenko) [inline drivers/i3c/device.c:140]:
```suggestion
 * Name: i3c_device_enable_ibi
```

File: drivers/i3c/device.c [hunk 2]
```diff
@@ -0,0 +1,346 @@
...
+  if (dev->desc)
+    {
+      *info = dev->desc->info;
+    }
+
+  i3c_bus_normaluse_unlock(dev->bus);
+}
+
+/****************************************************************************
+ * Name: i3c_device_disable_ibi()
```
Comment (pkarashchenko) [inline drivers/i3c/device.c:106]:
```suggestion
 * Name: i3c_device_disable_ibi
```

File: drivers/i3c/device.c [hunk 3]
```diff
@@ -0,0 +1,346 @@
...
+
+  i3c_bus_normaluse_lock(dev->bus);
+  ret = i3c_dev_do_priv_xfers_locked(dev->desc, xfers, nxfers);
+  i3c_bus_normaluse_unlock(dev->bus);
+
+  return ret;
+}
+
+/****************************************************************************
+ * Name: i3c_device_get_info()
```
Comment (pkarashchenko) [inline drivers/i3c/device.c:79]:
```suggestion
 * Name: i3c_device_get_info
```

File: drivers/i3c/device.c [hunk 4]
```diff
@@ -0,0 +1,346 @@
...
+ ****************************************************************************/
+
+#include "internals.h"
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: i3c_device_do_priv_xfers()
```
Comment (pkarashchenko) [inline drivers/i3c/device.c:32]:
```suggestion
 * Name: i3c_device_do_priv_xfers
```

File: drivers/i3c/device.c [hunk 5]
```diff
@@ -0,0 +1,346 @@
...
+      nxmutex_unlock(&dev->desc->ibi_lock);
+    }
+
+  i3c_bus_normaluse_unlock(dev->bus);
+
+  return ret;
+}
+
+/****************************************************************************
+ * Name: i3c_device_request_ibi()
```
Comment (pkarashchenko) [inline drivers/i3c/device.c:178]:
```suggestion
 * Name: i3c_device_request_ibi
```

File: drivers/i3c/device.c [hunk 6]
```diff
@@ -0,0 +1,346 @@
...
+      nxmutex_unlock(&dev->desc->ibi_lock);
+    }
+
+  i3c_bus_normaluse_unlock(dev->bus);
+
+  return ret;
+}
+
+/****************************************************************************
+ * Name: i3c_device_free_ibi()
```
Comment (pkarashchenko) [inline drivers/i3c/device.c:220]:
```suggestion
 * Name: i3c_device_free_ibi
```

File: drivers/i3c/device.c [hunk 7]
```diff
@@ -0,0 +1,346 @@
...
+      nxmutex_lock(&dev->desc->ibi_lock);
+      i3c_dev_free_ibi_locked(dev->desc);
+      nxmutex_unlock(&dev->desc->ibi_lock);
+    }
+
+  i3c_bus_normaluse_unlock(dev->bus);
+}
+
+/****************************************************************************
+ * Name: i3c_device_send_ccc_cmd()
```
Comment (pkarashchenko) [inline drivers/i3c/device.c:248]:
```suggestion
 * Name: i3c_device_send_ccc_cmd
```

File: drivers/i3c/device.c [hunk 8]
```diff
@@ -0,0 +1,346 @@
...
+
+  ret = i3c_master_send_ccc_cmd_locked(master, cmd);
+
+  i3c_bus_normaluse_unlock(dev->bus);
+
+  return ret;
+}
+
+/****************************************************************************
+ * Name: i3c_master_find_i3c_dev()
```
Comment (pkarashchenko) [inline drivers/i3c/device.c:284]:
```suggestion
 * Name: i3c_master_find_i3c_dev
```

File: drivers/i3c/i3c_driver.c [hunk 9]
```diff
@@ -0,0 +1,398 @@
...
+    {
+      return -ENOMEM;
+    }
+
+  priv->master = master;
+  nxmutex_init(&priv->lock);
+
+  /* Create the character device name */
+
+  snprintf(devname, DEVNAME_FMTLEN, DEVNAME_FMT, bus);
```
Comment (pkarashchenko) [inline drivers/i3c/i3c_driver.c:382]:
```suggestion
  snprintf(devname, sizeof(devname), DEVNAME_FMT, bus);
```

File: drivers/i3c/master.c [hunk 10]
```diff
@@ -0,0 +1,2282 @@
...
+  list_for_every_entry_safe(&master->bus.devs.i3c, i3cdev, i3ctmp,
+        struct i3c_dev_desc, common.node)
+    {
+      i3c_master_detach_i3c_dev(i3cdev);
+      kmm_free(i3cdev);
+    }
+}
+
+/****************************************************************************
+ * Name: i3c_master_bus_init()
```
Comment (pkarashchenko) [inline drivers/i3c/master.c:965]:
```suggestion
 * Name: i3c_master_bus_init
```

File: drivers/i3c/master.c [hunk 11]
```diff
@@ -0,0 +1,2282 @@
...
+{
+  char devname[12];
+  int i2c_bus_id = i3c_master_to_i2c_bus_number(master);
+
+  snprintf(devname, 12, "/dev/i2c%d", i2c_bus_id);
+  nx_unlink(devname);
+}
+
+/****************************************************************************
+ * Name: i3c_unregister_driver()
```
Comment (pkarashchenko) [inline drivers/i3c/master.c:1235]:
```suggestion
 * Name: i3c_unregister_driver
```

File: drivers/i3c/master.c [hunk 12]
```diff
@@ -0,0 +1,2282 @@
...
+ *   master - I3C master object.
+ *
+ ****************************************************************************/
+
+static void i2c_unregister_driver(FAR struct i3c_master_controller *master)
+{
+  char devname[12];
+  int i2c_bus_id = i3c_master_to_i2c_bus_number(master);
+
+  snprintf(devname, 12, "/dev/i2c%d", i2c_bus_id);
```
Comment (pkarashchenko) [inline drivers/i3c/master.c:1230]:
```suggestion
  snprintf(devname, sizeof(devname), "/dev/i2c%d", i2c_bus_id);
```

File: drivers/i3c/master.c [hunk 13]
```diff
@@ -0,0 +1,2282 @@
...
+       ops->free_ibi == NULL || ops->recycle_ibi_slot == NULL))
+    {
+      return -EINVAL;
+    }
+
+  return 0;
+}
+
+/****************************************************************************
+ * Name: i2c_unregister_driver()
```
Comment (pkarashchenko) [inline drivers/i3c/master.c:1215]:
```suggestion
 * Name: i2c_unregister_driver
```

File: drivers/i3c/master.c [hunk 14]
```diff
@@ -0,0 +1,2282 @@
...
+ * Input Parameters:
+ *   master - I3C master object.
+ *
+ ****************************************************************************/
+
+static void i3c_unregister_driver(FAR struct i3c_master_controller *master)
+{
+  char devname[12];
+
+  snprintf(devname, 12, "/dev/i3c%d", i3c_master_to_i3c_bus_number(master));
```
Comment (pkarashchenko) [inline drivers/i3c/master.c:1249]: sizeof

File: drivers/i3c/master.c [hunk 15]
```diff
@@ -0,0 +1,2282 @@
...
+ *
+ ****************************************************************************/
+
+void i3c_bus_normaluse_unlock(FAR struct i3c_bus *bus)
+{
+  nxmutex_unlock(&bus->lock);
+}
+
+/****************************************************************************
+ * Name: i3c_master_get_free_addr()
```
Comment (pkarashchenko) [inline drivers/i3c/master.c:1353]:
```suggestion
 * Name: i3c_master_get_free_addr
```

File: drivers/i3c/master.c [hunk 16]
```diff
@@ -0,0 +1,2282 @@
...
+
+  return pool;
+
+err_free_pool:
+  i3c_generic_ibi_free_pool(pool);
+  return ret;
+}
+
+/****************************************************************************
+ * Name: i3c_generic_ibi_get_free_slot()
```
Comment (pkarashchenko) [inline drivers/i3c/master.c:1920]:
```suggestion
 * Name: i3c_generic_ibi_get_free_slot
```

Comment (Donny9) [reply drivers/i3c/master.c]: Done~

File: drivers/i3c/device.c (hunk 1 above)
Comment (Donny9) [reply drivers/i3c/device.c]: Done~

File: drivers/i3c/device.c (hunk 2 above)
Comment (Donny9) [reply drivers/i3c/device.c]: Done~

File: drivers/i3c/device.c (hunk 3 above)
Comment (Donny9) [reply drivers/i3c/device.c]: Done~

File: drivers/i3c/device.c (hunk 4 above)
Comment (Donny9) [reply drivers/i3c/device.c]: Done~

File: drivers/i3c/device.c (hunk 5 above)
Comment (Donny9) [reply drivers/i3c/device.c]: Done~

File: drivers/i3c/master.c (hunk 12 above)
Comment (Donny9) [reply drivers/i3c/master.c]: Done~

File: drivers/i3c/master.c (hunk 13 above)
Comment (Donny9) [reply drivers/i3c/master.c]: Done~

File: drivers/i3c/master.c (hunk 14 above)
Comment (Donny9) [reply drivers/i3c/master.c]: Done~

File: drivers/i3c/master.c (hunk 15 above)
Comment (Donny9) [reply drivers/i3c/master.c]: Done~

**Pull request merged**